janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/addedrelations/SimpleAddedRelations.java

```diff
@@ -14,46 +14,74 @@
 
 package org.janusgraph.graphdb.transaction.addedrelations;
 
-import com.google.common.base.Predicate;
 import org.janusgraph.graphdb.internal.InternalRelation;
 
-import java.util.ArrayList;
+import com.carrotsearch.hppc.ObjectHashSet;
+import com.carrotsearch.hppc.ObjectSet;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Iterators;
+import javax.annotation.Nonnull;
+
+import java.util.AbstractCollection;
 import java.util.Collection;
-import java.util.List;
+import java.util.Collections;
+import java.util.Iterator;
 
 /**
  * @author Matthias Broecheler (me@matthiasb.com)
  */
 
-public class SimpleAddedRelations extends ArrayList<InternalRelation> implements AddedRelationsContainer {
-
+public class SimpleAddedRelations implements AddedRelationsContainer {
     private static final int INITIAL_ADDED_SIZE = 10;
+    private final ObjectSet<InternalRelation> container;
 
     public SimpleAddedRelations() {
-        super(INITIAL_ADDED_SIZE);
+        this.container = new ObjectHashSet<>(INITIAL_ADDED_SIZE);
     }
 
     @Override
     public boolean add(InternalRelation relation) {
-        return super.add(relation);
+        container.add(relation);
+        return true;
```

Comment:
Why it returns always `true`?
Should it be `return container.add(relation);`?

Reply:
This is set based container, and if element already exists it return false but client interpret it as addition failed